src/osd/OSD.cc

```diff
@@ -3492,6 +3507,11 @@ void OSD::_remove_heartbeat_peer(int n)
     q->second.con_front->mark_down();
   }
   heartbeat_peers.erase(q);
+  map<string, string> reporter_loc = osdmap->crush->get_full_location(n);
+  map<string, string>::iterator iter = reporter_loc.find(g_conf->mon_osd_reporter_subtree_level);
+  if (heartbeat_peers_by_subtree.count(iter->second))
```

Comment:
this needs to also clean up the osd.NNN format.  i'd make a simple helper to do this to avoid duplicating the code.  maybe even make it an OSDMap member?  OSDMap::get_osd_subtree(int osd, const string& level) ?